src/storage/store/engine.rs

```diff
@@ -224,48 +225,129 @@ impl ShardEngine {
+                    user_messages_count += 1;
+                }
+                Err(err) => {
+                    warn!(
+                        fid = msg.fid(),
+                        hash = msg.hex_hash(),
+                        "Error merging message: {:?}",
+                        err
+                    );
                 }
             }
+        }
 
-            for msg in &snap_txn.system_messages {
-                if let Some(onchain_event) = &msg.on_chain_event {
-                    self.onchain_event_store
-                        .merge_onchain_event(onchain_event.clone(), txn_batch)?;
+        for msg in &snap_txn.system_messages {
+            if let Some(onchain_event) = &msg.on_chain_event {
+                let event = self
+                    .onchain_event_store
+                    .merge_onchain_event(onchain_event.clone(), txn_batch);
 
-                    // TODO(aditi): Insert into the trie
+                match event {
+                    Ok(hub_event) => {
+                        self.update_trie(hub_event, txn_batch)?;
+                        system_messages_count += 1;
+                    }
+                    Err(err) => {
+                        warn!("Error merging onchain event: {:?}", err);
+                    }
                 }
             }
         }
 
-        let root1 = self.trie.root_hash()?;
+        info!(
+            fid = snap_txn.fid,
+            num_user_messages = total_user_messages,
+            num_system_messages = total_system_messages,
+            user_messages_merged = user_messages_count,
+            system_messages_merged = system_messages_count,
+            "Replayed transaction"
+        );
 
-        if &root1 != shard_root {
-            return Err(EngineError::HashMismatch);
-        }
+        // TODO: This should return the account root
+        Ok(())
+    }
 
+    fn merge_message(
+        &mut self,
+        msg: &message::Message,
+        txn_batch: &mut RocksDbTransactionBatch,
+    ) -> Result<hub_event::HubEvent, EngineError> {
+        let data = msg.data.as_ref().ok_or(EngineError::NoMessageData)?;
+        let mt = MessageType::try_from(data.r#type).or(Err(EngineError::InvalidMessageType))?;
+
+        let event = match mt {
+            MessageType::CastAdd => self
+                .cast_store
+                .merge(msg, txn_batch)
+                .map_err(EngineError::new_store_error(msg.hash.clone())),
+            unhandled_type => {
+                return Err(EngineError::UnsupportedMessageType(unhandled_type));
+            }
+        }?;
+
+        Ok(event)
+    }
+
+    fn update_trie(
+        &mut self,
+        event: hub_event::HubEvent,
+        txn_batch: &mut RocksDbTransactionBatch,
+    ) -> Result<(), EngineError> {
+        match event.body {
+            Some(hub_event::hub_event::Body::MergeMessageBody(merge)) => {
+                if let Some(msg) = merge.message {
```

Comment:
Do we expect merge.message to ever be None here? Should we return an error or panic in the missing case, or do you think it's ok or even normal to ignore?

Reply:
No, it's possible for it to be None. e.g. registering your fid for the first time will generate only an onchain event, or similarly, registering your fname will only have an fname transfer.